test/python/quantum_info/operators/test_operator.py

```diff
@@ -1050,6 +1051,113 @@ def test_from_circuit_mixed_reg_loose_bits_transpiled(self):
         result = Operator.from_circuit(tqc)
         self.assertTrue(Operator(circuit).equiv(result))
 
+    def test_apply_permutation_back(self):
+        """Test applying permutation to the operator,
+        where the operator is applied first and the permutation second."""
+        op = Operator(self.rand_matrix(64, 64))
+        pattern = [1, 2, 0, 3, 5, 4]
+
+        # Consider several methods of computing this operator and show
+        # they all lead to the same result.
+
+        # Compose the operator with the operator constructed from the
+        # permutation circuit.
+        op2 = op.copy()
+        perm_op = Operator(Permutation(6, pattern))
+        op2 &= perm_op
+
+        # Compose the operator with the operator constructed from the
+        # permutation gate.
+        op3 = op.copy()
+        perm_op = Operator(PermutationGate(pattern))
+        op3 &= perm_op
+
+        # Modify the operator using apply_permutation method.
+        op4 = op.copy()
+        op4 = op4.apply_permutation(pattern, front=False)
+
+        self.assertEqual(op2, op3)
+        self.assertEqual(op2, op4)
+
+    def test_apply_permutation_front(self):
+        """Test applying permutation to the operator,
+        where the permutation is applied first and the operator second"""
+        op = Operator(self.rand_matrix(64, 64))
+        pattern = [1, 2, 0, 3, 5, 4]
+
+        # Consider several methods of computing this operator and show
+        # they all lead to the same result.
+
+        # Compose the operator with the operator constructed from the
+        # permutation circuit.
+        op2 = op.copy()
+        perm_op = Operator(Permutation(6, pattern))
+        op2 = perm_op & op2
+
+        # Compose the operator with the operator constructed from the
+        # permutation gate.
+        op3 = op.copy()
+        perm_op = Operator(PermutationGate(pattern))
+        op3 = perm_op & op3
+
+        # Modify the operator using apply_permutation method.
+        op4 = op.copy()
+        op4 = op4.apply_permutation(pattern, front=True)
+
+        self.assertEqual(op2, op3)
+        self.assertEqual(op2, op4)
+
+    def test_reverse_qargs_as_apply_permutation(self):
+        """Test reversing qargs by pre- and post-composing with reversal
+        permutation.
+        """
+        perm = [3, 2, 1, 0]
+
+        for dims in [
+            (2, 3, 4, 5),
+            (5, 2, 4, 3),
+            (3, 5, 2, 4),
+            (5, 3, 4, 2),
+            (4, 5, 2, 3),
+            (4, 3, 2, 5),
+        ]:
+            op = Operator(
+                np.array(range(120 * 120)).reshape((120, 120)), input_dims=dims, output_dims=dims
+            )
+            op2 = op.reverse_qargs()
+            op3 = op.apply_permutation(perm, front=True).apply_permutation(perm, front=False)
+            self.assertEqual(op2, op3)
+
```

Comment:
I think we also need an explicit test that applying only a one-sided permutation to an `Operator` with heterogeneous tensored qudit spaces applies correctly (although I've got to say, I don't have a huge amount of intuition for what a clear test of that looks like).

Reply:
Done. Also added tests for applying the permutation [1, 0] to operator with dimensions (2, 3) x (2, 3), where things can be easily worked out explicitly.